Origin: http://selenic.com/hg/rev/e3f30068d2eb
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=783237
Description: sshpeer: more thorough shell quoting
 This fixes CVE-2014-9462.  Adapted to 1.6.4 by Javi Merino <vicho@debian.org>
Applied-Upstream: 3.2.4

--- a/mercurial/sshrepo.py
+++ b/mercurial/sshrepo.py
@@ -20,6 +20,14 @@ class remotelock(object):
         if self.repo:
             self.release()
 
+def _serverquote(s):
+    if not s:
+        return s
+    '''quote a string for the remote shell ... which we assume is sh'''
+    if re.match('[a-zA-Z0-9@%_+=:,./-]*$', s):
+        return s
+    return "'%s'" % s.replace("'", "'\\''")
+
 class sshrepository(repo.repository):
     def __init__(self, ui, path, create=0):
         self._url = path
@@ -37,7 +45,10 @@ class sshrepository(repo.repository):
         sshcmd = self.ui.config("ui", "ssh", "ssh")
         remotecmd = self.ui.config("ui", "remotecmd", "hg")
 
-        args = util.sshargs(sshcmd, self.host, self.user, self.port)
+        args = util.sshargs(sshcmd,
+                            _serverquote(self.host),
+                            _serverquote(self.user),
+                            _serverquote(self.port))
 
         if create:
             cmd = '%s %s "%s init %s"'
